Order parameter interface classes

jbair34 opened the pull request:
Closes #6279 . Needs to be approved by @dschwen. I am very paranoid about this going terribly wrong now...

jbair34 committed:
Created OPInterfaceBarrierMaterial #6279

added OPInterface to PhaseFieldApp.C #6279

made test for OPInterfacebarrier #6279

test for OPInterfacebarrier #6279

fixed branch issues hopefully #6279

got rid of tabs #6279

updated couple of errors #6279

fixed errors in repo

Fixed wrong maths

Fixed test

Fixed errors in OPInterfaceBarrier #6279

Fixed errors in OPInterfaceBarrier #6279

dschwen commented:
The tabs again. You can set Atom to a) make the whitespace visible, b) force the use of soft tabs (spaces)

jbair34 commented:
Sorry I do have in the settings of atom the soft tabs box checked. Where is the one that shows white spaces? I don't see it in the settings tab. I also tried to delete the only spot where i see red in the git log p thing but I guess it didn't work because it is still showing up...

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,84 @@
+    for (unsigned int j = i; j < _num_eta; ++j)
+    	for (unsigned int k = i; k < _num_eta; ++k)
+
+    {
```
indent and empty line (is this where the tab hides?)

permcody commented:
@jbair34 - Thanks! However now you have one commit by root (not you) and two little whitespace fixes by you. I suggest that you squash all three of these commits and take ownership of that commit by root.

permcody commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    for (unsigned int j = i; j < _num_eta; ++j)
+      for (unsigned int k = i; k < _num_eta; ++k)
+    {
```
Indentation is still not right here: Make sure the braces line up with the line above.

permcody commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+    _eta[i] = &coupledValue("etas", i);
+    for (unsigned int j = i; j < _num_eta; ++j)
+    {
+      _prop_d2g[i][j] =
```
This is an interesting line of code!

jbair34 replied:
Interesting as in wrong or just done oddly?

permcody replied:
Actually I don't know your physics well enough to say for certain and there isn't a comment here... Ignoring whitespace this is what you've written:

```
_prop_d2g[i][j] = _prop_d2g[j][i] = &declarePropertyDerivative<Real>(_function_name, eta_name[i], eta_name[j]);
```

So if you are intending to build a symmetric matrix this is correct, still probably worth a comment just to be sure, also the parenthesis aren't necessary. This is actually a single line of code.

jbair34 replied:
That is what I am trying to do. I got rid of the parenthesis.

permcody commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+
+  for (unsigned int i = 0; i < _num_eta; ++i)
+  {
+    _prop_dg[i]  = &declarePropertyDerivative<Real>(_function_name, eta_name[i]);
```
extra space before '='

jwpeterson commented:
> I suggest that you squash all three of these commits and take ownership of that commit by root.

After you've squashed everything down, change the author by doing:
```
git commit --amend --author=<Jacob Bair <jbair34@gmail.com>
```

permcody commented on modules/phase_field/include/materials/OPInterfaceBarrierMaterial.h:
```diff
@@ -0,0 +1,49 @@
+
+  /// order parameters
+  unsigned int _num_eta;
+  std::vector<VariableValue *> _eta;
```
This line needs to hold constant pointers due to a change we made in the framework recently.

```
std::vector<const VariableValue *> _eta;
```

jbair34 commented:
Thanks guys, Sorry for being a pain, seems like that happens too often when I try to pull something...

permcody commented:
No problem, I'm glad you're a good sport about all of our comments. However I have to comment on your statement about "pull". We're no longer recommending that people ever use that command *ever*. We've updated our docs to reflect what we believe to be a more logical set of git commands to use. (i.e. fetch and rebase). You may want to take a look at our "contributing" slides at some point.

jbair34 commented:
Ok, I will take a look. I meant make a pull request but it just so happens that I have still been using git pull so that is a good thing to know. Thanks.

jbair34 commented:
Holy Hell, I swear this compiled yesterday! But apparently something has happened since then that is causing it to give me a ton of errors.. I will try to get to this in a while but have some meetings now...

jbair34 commented:
Well I apparently did something that completely screwed up my repo when I switched to fetch from pull but I re cloned everything and it all compiles again... But that line where you (@permcody)  said I didn't need the parenthesis doesn't compile without them so I left them in... I was assuming you meant () as parenthesis but did you maybe mean some of the square brackets or something?

permcody commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+    for (unsigned int j = i; j < _num_eta; ++j)
+      for (unsigned int k = i; k < _num_eta; ++k)
+      {
+      switch (_g_order)
```
Do you plan on adding additional cases later? Just curious

jbair34 replied:
Not for my purposes... I guess we could get rid of that I just did it in case someone wanted to do something different I guess...

permcody commented:
@dschwen - please review

jbair34 commented:
I noticed that some of the math didn't do what I wanted it too because I didn't want the possibility of having i == j or k because this material should only affect areas where 3 or more order parameters are present. Hopefully I have updated it correctly to reflect those changes...

aeslaughter commented:
@dschwen @jbair34 What is the status of this, it looks like from the comments that this still needs a final review.

jbair34 commented:
Yes it does, Daniel has been pretty busy I think... He just needs to give the ok on my math part of it I think.

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+  params.addParam<std::string>("function_name", "g", "actual name for g(eta_i)");
+  MooseEnum h_order("SIMPLE=0", "SIMPLE");
+  params.addParam<MooseEnum>("g_order", h_order, "Polynomial order of the switching function h(eta)");
+  params.addRequiredCoupledVar("etas", "eta_i order parameters, one for each h");
```
fix doc string. there is no h!

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+  InputParameters params = validParams<Material>();
+  params.addParam<std::string>("function_name", "g", "actual name for g(eta_i)");
+  MooseEnum h_order("SIMPLE=0", "SIMPLE");
+  params.addParam<MooseEnum>("g_order", h_order, "Polynomial order of the switching function h(eta)");
```
g_order or h_order. This should be consistent, or - preferably - removed, since there is only one option now.

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+    _prop_d2g(_num_eta)
+
+{
+
```
remove the two empty lines

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    for (unsigned int j = 0; j < _num_eta; ++j)
+    {
+      if (i == j) continue;
```
I'd rather have the user pass in a matrix of prefactors

dschwen replied:
This construct is not very elegant. It would be more efficient to just loop like this:

```
for (unsigned int i = 0; i < _num_eta; ++i)
  for (unsigned int j = 0; j < i; ++j)
    for (unsigned int k = 0; k < j; ++k)
    {
      ...
    }
```

This avoid looping unnecessarily over equivalent permutations if `i, j, k` and the `i=j | j=k | i=k` cases.

dschwen replied:
Wait...

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+        switch (_g_order)
+        {
+          case 0: // SIMPLE
+            _prop_g[_qp]         +=  (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp] * (*_eta[k])[_qp] * (*_eta[k])[_qp];
```
Yeah, a matrix (rank-three tensor) is too cumbersome, but I suggest adding a prefector that can be controlled by a parameter.

jbair34 replied:
So I specifically didn't add a prefactor because the MultiBarrierFunctionMaterial does not have one and the prefactor is controlled by the DerivativeMultiphaseMaterial anyways... Do you want me to add one? I thought it might be confusing to have two places to add the prefactor.

dschwen replied:
You are right! That is fine. I forgot about that external prefactor.

On Tue, Mar 15, 2016, 7:48 AM jbair34 <notifications@github.com> wrote:

> In modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C
> <https://github.com/idaholab/moose/pull/6460#discussion_r56165145>:
>
> > +    for (unsigned int j = i; j < _num_eta; ++j)
> > +      (*_prop_d2g[i][j])[_qp] = 0.0;
> > +  }
> > +
> > +  // Create Interface barrier preventing interfaces involving more than two order parameters
> > +  for (unsigned int i = 0; i < _num_eta; ++i)
> > +    for (unsigned int j = 0; j < _num_eta; ++j)
> > +    {
> > +      if (i == j) continue;
> > +      for (unsigned int k = 0; k < _num_eta; ++k)
> > +      {
> > +        if ((i == k) || (j == k)) continue;
> > +        switch (_g_order)
> > +        {
> > +          case 0: // SIMPLE
> > +            _prop_g[_qp]         +=  (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp] * (*_eta[k])[_qp] * (*_eta[k])[_qp];
>
> So I specifically didn't add a prefactor because the
> MultiBarrierFunctionMaterial does not have one and the prefactor is
> controlled by the DerivativeMultiphaseMaterial anyways... Do you want me to
> add one? I thought it might be confusing to have two places to add the
> prefactor.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/idaholab/moose/pull/6460/files#r56165145
>

dschwen commented on modules/phase_field/include/materials/OPInterfaceBarrierMaterial.h:
```diff
@@ -0,0 +1,49 @@
+InputParameters validParams<OPInterfaceBarrierMaterial>();
+
+/**
+ * OPInterfaceBarrierMaterial is a constraint kernel that acts on all of the eta_i variables to
```
No, it is not a kernel. It is a Free Energy penalty contribution (and a Material)

dschwen commented on modules/phase_field/include/materials/OPInterfaceBarrierMaterial.h:
```diff
@@ -0,0 +1,49 @@
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef OPINTERFACEBARRIER_H
+#define OPINTERFACEBARRIER_H
```
include guards should be names after the class name. I.e. `OPINTERFACEBARRIERMATERIAL_H`

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+template<>
+InputParameters validParams<OPInterfaceBarrierMaterial>()
+{
+  InputParameters params = validParams<Material>();
```
Just a suggestion, but it would be helpful to add a class description here using
```
params.addClassDescriptipon("Free Energy contribution that penalizes more than three order parameters being non-zero");
```
```

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,83 @@
+{
+
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    _prop_d2g[i].resize(_num_eta);
```
This loop could be combined with the loop directly below.

dschwen commented on modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C:
```diff
@@ -0,0 +1,68 @@
+      for (unsigned int k = 0; k < j; ++k)
+      {
+            _prop_g[_qp]         +=  (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp] * (*_eta[k])[_qp] * (*_eta[k])[_qp];
+            (*_prop_dg[i])[_qp]  +=  2 * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp] * (*_eta[k])[_qp] * (*_eta[k])[_qp];
```
This derivative (and teh second order derivative) are now not correct anymore. This needs two more lines with `(*_prop_dg[j])[_qp]` and `(*_prop_dg[k])[_qp]`.

jbair34 replied:
shouldn't (_prop_dg[i])[_qp] cover all first derivatives (dF/deta_i) as it cycles through i's without needing anything for j and k? Also the math wasn't changed at all and these derivative should still be correct for dF/deta_i and d^2F/(deta_i*deta_j) with the same logic as before that cycling through those combinations should cover all off-diagonal second derivatives... However, the second derivative should actually have a case for d^2F/deta_i^2 so I will add that if you want me to...

dschwen replied:
Look at each different `i` and think about how many `j` and `k` contribute to its term...

jbair34 replied:
Oh So are you saying because we got rid of those continues and set up this new for loop where k < j < i we need derivatives for the j and k as well? Assuming that is what the new for loop does...?

dschwen replied:
Yes, exactly.

dschwen commented on modules/phase_field/include/materials/OPInterfaceBarrierMaterial.h:
```diff
@@ -0,0 +1,47 @@
+#include "DerivativeMaterialInterface.h"
+
+// Forward Declarations
+class OPInterfaceBarrierMaterial;
```
I would like to come up with a better name for this. We already have two materials providing "order parameter interface" barrier contributions. This material is to suppress third phases. How about something like `ThirdPhaseSuppressionMaterial`?

dschwen commented:
I would like to put this on ice until I finished addressing #6584, which involves a rewrite of a related class. I have to figure out some of the math needed here for that PR.

jbair34 commented:
@dschwen I did do the changes you suggested but go ahead and finish the crossterm stuff before you look at it like you said. I just wanted to push the fixes...

jbair34 commented:
@dschwen do I need second derivatives for [j][i] and [i][j] or do the ones I have in there cover it?

jbair34 commented:
Did I do something weird with commits again? There seems to be some that aren't mine here...

jbair34 committed:
Moved OPInterface... to ThirdPhaseSuppressionMaterial and updated derivatives

added shortened names for etas to ThirdPhaseSuppresionMaterial

Removed whitespace ThirdPhaseSuppressionMaterial

permcody commented:
You just need to rebase your commits. However you also have a conflict to resolve now too.

jbair34 commented:
Where can I find the conflict? It didn't show up when I committed last time..? Also how do I rebase a commit? Just rebase now and do a new commit?

jbair34 commented:
Well I think the conflict was because of the need to rebase... At least it says no conflicts now and that is all I changed...

jbair34 commented:
@dschwen can you take a look at this when you get some time?

dschwen commented:
I don't think the SmoothCircleIC based test is particularly useful anymore. I switched the other barrier materials to 1D and function ICs. This one could be a 2D setup with two OPs crossing over on the x-axis and a third OP coming up along the y-axis.

jbair34 commented:
Like three bounding boxes with a triple junction?

dschwen commented:
Just Use FunctionICs in x and y direction
box from 0..1 in bot x and y direction
eta1 = x
eta2 = 1-x
eta3 = y

On Tue, Mar 22, 2016 at 9:48 AM jbair34 <notifications@github.com> wrote:

> Like three bounding boxes with a triple junction?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6460#issuecomment-199875134>
>

jbair34 commented:
@dschwen have you had a chance to look at this with the new test?

jbair34 committed:
Changed test to use functionICs

fixed indentation issues

dschwen commented:
The new test looks fine. Let me look over the code again.

dschwen commented on modules/phase_field/src/materials/ThirdPhaseSuppressionMaterial.C:
```diff
@@ -0,0 +1,79 @@
+  std::vector<std::string> eta_name(_num_eta);
+  for (unsigned int i = 0; i < _num_eta; ++i)
+  {
+      _prop_d2g[i].resize(_num_eta);
```
indentation is off

dschwen commented on modules/phase_field/src/materials/ThirdPhaseSuppressionMaterial.C:
```diff
@@ -0,0 +1,79 @@
+        const Real nj = (*_eta[j])[_qp];
+        const Real nk = (*_eta[k])[_qp];
+
+            _prop_g[_qp]         +=  ni * ni * nj * nj * nk * nk;
```
indentation

dschwen commented:
Before this is merged I'd like to see those derivatives verified.

permcody commented:
You updated a previous comment so I didn't get the email. You think this all checks out now?

jbair34 commented:
I still need to fix those indentations but I am home this afternoon so I will have to do that tomorrow morning...

jbair34 commented:
Indents were updated and tests pass. Let me know if anything else needs to be done by me...
